src/common/ddax_deep_flush.c

```diff
@@ -0,0 +1,212 @@
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+/*
+ * ddaxe_deep_flush.c -- Internal utility functions for flushing
+ * a memory range residing on a DAX device.
+ * Currently only used on Linux.
+ */
+
+#define _GNU_SOURCE
+
+
+#include <errno.h>
+#include <fcntl.h>
+#include <sys/stat.h>
+#include <unistd.h>
+
+#include "os.h"
+#include "file.h"
+#include "out.h"
+#include "ddax_deep_flush.h"
+#include "libpmem.h"
+
+#define DAX_REGION_ID_LEN 4
+#define MAX_UNIQUE_REGIONS 1024
+
+/*
+ * ddax_region_find - looks for region number for specified dev_id
+ */
+int
+ddax_region_find(os_dev_t dev_id)
+{
+
+	LOG(2, "ddax_region_find %lu", (unsigned long)dev_id);
+#ifndef _WIN32
+	int dax_reg_id_fd;
+	char dax_region_path[PATH_MAX];
+	char reg_id[DAX_REGION_ID_LEN + 1];
+
+	snprintf(dax_region_path, PATH_MAX,
+		"/sys/dev/char/%u:%u/device/dax_region/id",
+		major(dev_id), minor(dev_id));
+
+	if ((dax_reg_id_fd = os_open(dax_region_path, O_RDONLY)) < 0) {
+		ERR("open(\"%s\", O_PATH", dax_region_path);
+		return -1;
+	}
+
+	ssize_t len = pread(dax_reg_id_fd, reg_id, sizeof(int), 0);
+
+	if (len <= 1 || len > DAX_REGION_ID_LEN + 1) {
+		ERR("pread(%d, %p, %lu, 0)", dax_reg_id_fd, reg_id,
+		sizeof(int));
+		goto err;
+	}
+
+	os_close(dax_reg_id_fd);
+	return atoi(reg_id);
+
+err:
+	os_close(dax_reg_id_fd);
+	return -1;
+#else
+	return -1;
+#endif
+}
+
+/*
+ * ddax_deep_flush - check if deep flush request is for pmem or other
+ */
+int
+ddax_deep_flush(const void *addr, size_t len, struct pool_set *set,
+unsigned replica_id)
+{
+	LOG(2, "ddax_deep_flush addr %p len %zu set %p replica_id %u",
+		addr, len, set, replica_id);
+
+	if (set == NULL) {
+		if (pmem_deep_flush(addr, len)) {
+			ERR("!pmem_deep_flush(%p, %lu)", addr, len);
+			return -1;
+		}
+	} else {
+		if (ddax_replica_deep_flush(addr, len, set, replica_id)) {
+			ERR("!ddax_replica_deep_flush(%p, %lu, %p, %u)",
+				addr, len, set, replica_id);
+			return -1;
+		}
+	}
+
+	return 0;
+}
+
+/*
+ * ddax_replica_deep_flush -- perform deep flush on replica's parts from range
+ * for dev dax write to deep_flush, otherwise run msync
+ */
+int
+ddax_replica_deep_flush(const void *addr, size_t len,
+struct pool_set *set, unsigned replica_id)
+{
+	LOG(2, "ddax_replica_deep_flush addr %p len %zu set %p replica_id %u",
+		addr, len, set, replica_id);
+
+	int unique[MAX_UNIQUE_REGIONS];
+	int flushed;
+	int u = 0;
+
+	struct pool_replica *rep = set->replica[replica_id];
+	uintptr_t rep_addr = (uintptr_t)rep->part[0].addr;
+	uintptr_t rep_end = rep_addr + rep->repsize;
+	uintptr_t end = (uintptr_t)addr + len;
+
+	ASSERT((uintptr_t)addr >= rep_addr);
+	ASSERT(end <= rep_end);
+
+	for (int u = 0; u < MAX_UNIQUE_REGIONS; u++) {
+		unique[u] = -1;
```

Comment:
you can try `= {[0 ... MAX_UNIQUE_REGIONS-1] = -1};`

but I'm not sure if this is a valid C or GNU C?